clang/lib/CIR/CodeGen/CIRGenTypes.cpp

```diff
@@ -281,3 +281,33 @@ bool CIRGenTypes::isZeroInitializable(clang::QualType t) {
 
   return true;
 }
+
+const CIRGenFunctionInfo &CIRGenTypes::arrangeCIRFunctionInfo() {
+  // Lookup or create unique function info.
+  llvm::FoldingSetNodeID id;
+  CIRGenFunctionInfo::Profile(id);
+
+  void *insertPos = nullptr;
+  CIRGenFunctionInfo *fi = functionInfos.FindNodeOrInsertPos(id, insertPos);
+  if (fi)
+    return *fi;
```

Comment:
I don't like the idea that we're returning a reference to a pointer we got back from the folding set. Do we have any guarantees about the lifetime of that pointer?

Reply:
A folding set is basically a separate chaining hash table, and the addresses of its elements are stable once the element is inserted. Besides, in the classic codegen, we never erase elements from `functionInfos`. Thus the returned reference is valid throughout CIRGen.